File: windows/Tokendial.Core/Install/InstallState.cs
namespace Tokendial.Core.Install;

/// <summary>How far a provider's tool has come: the three steps the settings row draws, plus the connected end state.</summary>
public enum InstallState
{
    NotInstalled,
    Installed,
    SignedIn,
    Connected
}

public static class InstallProgress
{
    public static InstallState Of(bool installed, bool signedIn, bool connected)
    {
        if (signedIn) return connected ? InstallState.Connected : InstallState.SignedIn;
        return installed ? InstallState.Installed : InstallState.NotInstalled;
    }
}

File: macos/TokendialCore/Sources/TokendialCore/Install/InstallState.swift
import Foundation

/// How far a tool has got. The order carries meaning: the step strip asks whether the state has reached a
/// step, so the cases must stay in this sequence.
public enum InstallState: Int, Comparable, Sendable {
    case notInstalled
    case installed
    case signedIn
    case connected

    public static func < (a: InstallState, b: InstallState) -> Bool { a.rawValue < b.rawValue }

    /// Connected is only read once the tool is signed in: ticking a provider that cannot be read yet is not
    /// progress, it is a preference waiting for a credential.
    public static func of(installed: Bool, signedIn: Bool, connected: Bool) -> InstallState {
        guard signedIn else { return installed ? .installed : .notInstalled }
        return connected ? .connected : .signedIn
    }
}
